### PR TITLE
Updates caching docs to be all about the new style of caching

### PR DESCRIPTION
**Pull request by cicdw**
Note that some changes depend on #14164 

My goal here was to give a practical orientation to caching without mentioning transactions, as those will be more advanced

### CODE REVIEW

File: docs/3.0rc/develop/write-tasks/index.mdx [hunk 1]
```diff
@@ -483,148 +483,223 @@ def my_flow():
...
+and ensure that your pipelines are idempotent when retrying them due to unexpected failure. 
 
-Define a task that is cached based on its inputs by using the Prefect 
-[`task_input_hash`](/3.0rc/api-ref/prefect/tasks/#prefect.tasks.task_input_hash).
+By default Prefect's caching logic is based on the following attributes of a task invocation:
+- the inputs provided to the task
+- the code definition of the task
+- the prevailing flow run ID, or if executed autonomously, the prevailing task run ID
+
+All of these are hashed to compute the task's _cache key_.
```
Comment (jlowin) [inline docs/3.0rc/develop/write-tasks/index.mdx:496]: I think it is important to acknowledge the expected consequence is that a task called twice in the same flow with the same arguments will not run twice but reuse its cached results -- inter-flow idempotency.

File: docs/3.0rc/develop/write-tasks/index.mdx [hunk 2]
```diff
@@ -483,148 +483,223 @@ def my_flow():
...
-To determine whether a task run should retrieve a cached state, we use "cache keys". A cache key is a 
-string value that indicates if one run should be considered identical to another. When a task run with a 
-cache key finishes, we attach that cache key to the state. When each task run starts, Prefect checks for 
-states with a matching cache key. If a state with an identical key is found, Prefect uses the cached 
-state instead of running the task again.
+Prefect comes prepackaged with a few common cache policies:
+- `DEFAULT`: this cache policy uses the task's inputs, it's code definition, as well as the prevailing flow run ID
+to compute the task's cache key.
+- `INPUTS`: this cache policy uses _only_ the task's inputs to compute the cache key.
+- `TASKDEF`: this cache policy uses _only_ the task's code definition to compute the cache key.
```
Comment (jlowin) [inline docs/3.0rc/develop/write-tasks/index.mdx:534]: FWIW, all the other values here are self documenting but this one really isn't clear unless you read this doc. 

Suggestions:

`TASK_CODE`
`TASK_DEF` (maybe)
`TASK_SOURCE`

File: docs/3.0rc/develop/write-tasks/index.mdx [hunk 3]
```diff
@@ -483,148 +483,223 @@ def my_flow():
...
 
-### Caching example
+my_stateful_task() # sleeps
+my_stateful_task() # does not sleep
+```
 
-In this example, until the `cache_expiration` time ends, as long as the input to `hello_task()` remains 
-the same when it is called, the cached return value is returned. In this situation the task is not rerun. 
-However, if the input argument value changes, `hello_task()` runs using the new input.
+This task will run once and only once until the underlying code is altered:
```
Comment (jlowin) [inline docs/3.0rc/develop/write-tasks/index.mdx:558]:
```suggestion
No matter how many flows call it, this task will run once and only once until its underlying code is altered:
```

**Review by jlowin — APPROVED** [review]

File: docs/3.0rc/develop/write-tasks/index.mdx (hunk 1 above)
Comment (cicdw) [reply docs/3.0rc/develop/write-tasks/index.mdx:496]: Good call, will do

File: docs/3.0rc/develop/write-tasks/index.mdx (hunk 2 above)
Comment (cicdw) [reply docs/3.0rc/develop/write-tasks/index.mdx:534]: Love it; `TASK_SOURCE` sounds best to me but it's a bit formal. Will need to think about which to use but I'll change it either way

Comment (cicdw) [reply docs/3.0rc/develop/write-tasks/index.mdx:534]: source actually hints at how I get this (`getsource()`) so I think it's the best

Comment (cicdw) [reply docs/3.0rc/develop/write-tasks/index.mdx:534]: Made this change in my code companion PR to this: https://github.com/PrefectHQ/prefect/pull/14164/commits/e8c996d0d5a391bb8f11118c5cd59a07665c494d

File: docs/3.0rc/develop/write-tasks/index.mdx (hunk 1 above)
Comment (cicdw) [reply docs/3.0rc/develop/write-tasks/index.mdx]: @jlowin updated

**Pull request merged**